crates/pathfinder/src/rpc/v02/method/get_proof.rs

```diff
@@ -0,0 +1,272 @@
+    pub contract_address: ContractAddress,
+    pub key: StorageAddress,
+    pub block_id: BlockId,
+}
+
+crate::rpc::error::generate_rpc_error_subset!(
+    GetStorageProofError: ContractNotFound,
+    BlockNotFound
+);
+
+/// TODO: description
+pub async fn get_proof(
+    context: RpcContext,
+    input: GetStorageProofInput,
+) -> Result<Vec<ProofNode>, GetStorageProofError> {
+    let block_id = match input.block_id {
+        BlockId::Hash(hash) => hash.into(),
+        BlockId::Number(number) => number.into(),
+        BlockId::Latest => StarknetBlocksBlockId::Latest,
+        BlockId::Pending => {
+            // TODO: what to do with pending blockId?
+            match context
+                .pending_data
+                .ok_or_else(|| anyhow!("Pending data not supported in this configuration"))?
+                .state_update()
+                .await
+            {
+                Some(update) => {
+                    let pending_value = update
+                        .state_diff
+                        .storage_diffs
+                        .get(&input.contract_address)
+                        .and_then(|storage| {
+                            storage.iter().find_map(|update| {
+                                (update.key == input.key).then_some(update.value)
+                            })
+                        });
+
+                    // match pending_value {
+                    //     Some(value) => return Ok(value),
+                    //     None => StarknetBlocksBlockId::Latest,
+                    // }
+                    StarknetBlocksBlockId::Latest
+                }
+                None => StarknetBlocksBlockId::Latest,
+            }
+        }
+    };
+
+    let storage = context.storage.clone();
+    let span = tracing::Span::current();
+
+    let jh = tokio::task::spawn_blocking(move || {
+        let _g = span.enter();
+        let mut db = storage
+            .connection()
+            .context("Opening database connection")?;
+
+        let tx = db.transaction().context("Creating database transaction")?;
+
+        // Use internal error to indicate that the process of querying for a particular block failed,
+        // which is not the same as being sure that the block is not in the db.
+        let global_root = StarknetBlocksTable::get_root(&tx, block_id)
+            .context("Get global root for block")?
+            // Since the db query succeeded in execution, we can now report if the block hash was indeed not found
+            // by using a dedicated error code from the RPC API spec
+            .ok_or(GetStorageProofError::BlockNotFound)?;
+
+        let global_state_tree =
+            GlobalStateTree::load(&tx, global_root).context("Global state tree")?;
+
+        let contract_state_hash = global_state_tree
+            .get(input.contract_address)
+            .context("Get contract state hash from global state tree")?
+            .ok_or(GetStorageProofError::ContractNotFound)?;
```

Comment:
We need to use the merkle proof here as well. And if the contract does not exist, we should still send the proof back, so they can confirm that is true.

Reply:
Indeed, very good idea. I don't know what I had not thought about it, makes total sense :p